### ISSUE TITLE
Wrong error score when using binarylog or logistic as objective function in vl_nnloss

### ISSUE THREAD

**Issue by sunset123** [vlfeat/matconvnet#499]
Hi,
I have images that have multiple attributes and I want to train a CNN to predict for each image the presence or absence of a particular attribute. I am testing the vl_nnloss for this and I have the following issues

1) using logistic objective function in vl_nnloss produce wrong error score even when X=c. For example the following code will result in Y=12.5 which is wrong because X=c.

c=[ones(10,1),ones(10,1)*-1;ones(10,1)*-1,ones(10,1)];
c=reshape(c,1,1,2,20);
X=c;
Y = vl_nnloss(X,c);

2) using binarylog works with previous example but results in infinity when the following code is used:
c=[];
X=[];
c(1,1,:,1)=[1 0];
X(1,1,:,1)=[0 1];
Y = vl_nnloss(X,c);

Can you please help me to fix this issue.

Many thanks,
Heba

**Comment by albanie** [comment]
Matconvnet uses the convention that labels marked as 0 should be skipped - they do not contribute to the loss.  

Note that the loss produced by the logistic objective function defined in `vl_nnloss` will still be non zero.  This is because logistic loss assigns a penalty to every term, even those that are classified correctly.

**Comment by sunset123** [comment]
Thanks. So is it ok for the logistic log to have high loss even when prediction and scores are identical?

I used non zeros (-1, 1) for the binarylog but it still producing inf. Also it is a bit confusing because in vl_nnloss documentation it says that X should be a probability in the range [0,1] when using bianarylog as following:

%   Binary log loss:: `binarylog`
%     L(x,c) = - log(c(x-0.5) + 0.5). x is assumed to be the
%     probability that the attribute is active (c=+1). Hence x must be
%     a number in the range [0,1]. This is the binary version of the
%     `log` loss.

Ultimately, I would like to use and I am not sure how to avoid the inf scores?

Thanks

**Comment by jotaf98** [comment]
Yes, that's expected behavior. The log loss has singularities at the extremes.

Although they're not exactly the same equations, you can read more about it here:
http://www.r-bloggers.com/making-sense-of-logarithmic-loss/

**Comment by sunset123** [comment]
Thanks. So if I want to specify the probability that a given image has an attribute i when using binarylog, I should use a number in (0,1] range. Is this right?

Also do negative scores (e.g. -0.5) imply the probability of absence of attribute?
